File: algorithms/java/src/test/java/leetcode/LC378KthSmallestElementInASortedMatrix.java
package leetcode;

import org.junit.Test;

import java.util.Collections;
import java.util.PriorityQueue;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/
 */
public class LC378KthSmallestElementInASortedMatrix {
    public class Solution {
        public int kthSmallest(int[][] matrix, int k) {
            PriorityQueue<Integer> heap = new PriorityQueue(k, Collections.reverseOrder());
            for (int[] row : matrix) {
                for (int cell : row) {
                    heap.add(cell);
                    if (heap.size() > k) {
                        heap.poll();
                    }
                }
            }
            return heap.poll();
        }
    }

    @Test
    public void test_example() throws Exception {
        int[][] matrix = {
                {1, 5, 9},
                {10, 11, 13},
                {12, 13, 15}
        };
        assertEquals(13, new Solution().kthSmallest(matrix, 8));
    }
}

File: algorithms/code/leetcode/lc378_kth_smallest_element_in_a_sorted_matrix/lc378_kth_smallest_element_in_a_sorted_matrix.py
#!/usr/bin/env python3
# https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix/

import heapq
import unittest
from typing import Any


class Solution:
    class Item:
        def __init__(self, value: int) -> None:
            self.value = value

        def __lt__(self, other: Any) -> bool:
            return other.value < self.value

    def kthSmallest(self, matrix: list[list[int]], k: int) -> int:
        heap: list[Solution.Item] = []
        for row in matrix:
            for cell in row:
                heapq.heappush(heap, Solution.Item(cell))
                if len(heap) > k:
                    heapq.heappop(heap)
        return heapq.heappop(heap).value


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        matrix = [[1, 5, 9], [10, 11, 13], [12, 13, 15]]
        self.assertEqual(13, Solution().kthSmallest(matrix, 8))
